### DISCUSSION TITLE
Transition issues with dynamic slide rendering (without breaking other apis)

### DISCUSSION THREAD

**Discussion by remy90** [davidjerleke/embla-carousel#467] [Help] [answered]
Hi there, I'm having issues with rendering dynamically fetched content. The client knows the total number of questions but not the next question until an answer has been selected. The following works with no transition issues:

```typescript
const scrollNext = useCallback(async () => {
    if (emblaApi && !isFetching) {
      const nextIndex = selectedIndex + 1

      if (nextIndex === totalQuestions) return

        emblaApi.scrollNext()
        setIsFetching(true)
        const nextQuestion = await fetchQuestions(surveyType, journeyId, authString)
        if (nextQuestion?.questionDetails) {
          setQuestions([...questions, nextQuestion.questionDetails])
        }
        setSelectedIndex(selectedIndex + 1)
        setNextBtnEnabled(selectedIndex < totalQuestions)
        setPrevBtnEnabled(selectedIndex > 0)
        setIsFetching(false)
      }
  }, [...])
```

However it breaks `canScrollNext()` and probably a lot of other things. I've used [arrows and dots](https://www.embla-carousel.com/examples/static/#arrows--dots) as a starting template because we have forward and back buttons. I made good progress with a refactor back towards arrows and dots and used `emblaApi.on('settle', () => {...})` to wait for the transition to complete before setting state but it takes too long and we're introducing a skeleton for the transition rather than for network speeds.

I've looked at your suggestion towards [infinite scrolling](https://github.com/davidjerleke/embla-carousel/issues/103) but it seems like there's a hefty overhead, considering the above works.

Do you have any suggestions other than using the infinite scroll as a template in this use case? My main additional requirement to the arrows and dots example is:
```typescript
const nextQuestion = await fetchQuestions(surveyType, journeyId, authString)
        if (nextQuestion?.questionDetails) {
          setQuestions([...questions, nextQuestion.questionDetails]) // where setQuestions is useState<QuestionDetails[]>([questionSets])
        }
```

**Comment by davidjerleke** [comment]
Hi,

If I understand you correctly, you want to dynamically add slides after fetching them. The [latest release](https://github.com/davidjerleke/embla-carousel/releases/tag/v8.0.0-rc01) introduces the `watchSlides` option where you can intercept the Embla auto re-init behavior when slides are added/removed, and add your own logic before it re-initializes. 

Have you checked that out?

Best,
David

**Reply by remy90** [reply]
I've actually just been playing about with that exact example for the past hour 😅

**Reply by davidjerleke** [reply]
I'm not sure what problem you're experiencing?

**Reply by remy90** [reply]
Are you suggesting I should use watchSlides with arrows and dots to accomplish this? I think there's an example I could benefit from as I don't have a strong grasp of what it entails outside of the in depth example in infinite scrolls. Perhaps I've created an issue where there isn't one. I'm not currently using `canScrollNext()`, but with my implementation listed above, it's always false

**Reply by davidjerleke** [reply]
When do you want Embla to pick up and include the fetched and injected slides (meaning that the user should be able to scroll to them)? At what point?

**Reply by remy90** [reply]
At present, it's when they click the next button. I should add - totalQuestions updates per `fetchQuestions` fetch, so we know there are more questions ahead even if we don't know what that question is. I've updated the above code snippet to exclude the if block around `scrollNext`. The issue appears to be a race condition between the scrollNext() and the re-render triggered by setQuestions

**Reply by davidjerleke** [reply]
If you know how many questions there are in advance, and just clicking next should proceed to the next slide, couldn't you add empty slides for all questions and populate them as soon as they are in viewport? Either using [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) or the `emblaApi.slidesInView()`?

**Reply by davidjerleke** [reply]
Or are you dynamically fetching question content?

**Reply by remy90** [reply]
Dynamically fetching question content. I have a "Loading..." backup but it shouldn't be required. Currently, it's only text that's being loaded, so it's very quick. When images are eventually loaded, there will be more justification for the Loading skeleton

**Reply by davidjerleke** [reply]
I think I need to see a CodeSandbox with your setup so far. There are too many pieces missing for me to understand what you want to achieve.

**Reply by remy90** [reply]
Sorry for the delay, that probably took me a lot longer than it should have:
https://codesandbox.io/s/embla-carousel-arrow-dots-react-forked-zfw5c5?file=/src/js/EmblaCarousel.tsx
What you'll notice is the lack of transition when you click `scrollNext` the first time. The second time, the third question doesn't load at all (not sure why but there's a fallback message).

You can think of `SLIDES` as my mockApi, where each array item is drip fed to the client.

The use case for this is (eventual) conditional logic on the server for the next question (i.e. if Yes then Q3 else Q4) and large scale distrust of the client.

If I use `select.on('settle', () => {...}`, it loads once the transition is complete, but the response of `SLIDES` is similar to the network response speed currently, so we're actually waiting for the settle transition, rather than the data.

**Reply by davidjerleke** [reply]
Thanks @remy90. I think I need one more thing. A specification. How is it supposed to work? 

- Should the user be able to scroll to the next slide before the content of that slide has finished loading?
- Or should the carousel next button be disabled until the next slide content has finished loading?

**Reply by remy90** [reply]
Great question. On current implementations, the fetch happens on the next slide click. There could be future iterations where a selected answer triggers the load before the user clicks next but I think the next button shouldn't be disabled, so:

The next button should always be enabled when we know there's another question. Ideally, a loading screen is only displayed because of large data fetches or slow network requests, so the first option seems best IMO

I can create another sandbox with the code at the top to show how it works without issues relating to content or scrolling if that's helpful?

**Reply by remy90** [reply]
Any further thoughts David? I've also encountered some issues with using on settle that appears to run havoc on my unit tests so a work around would be much appreciated!

**Reply by davidjerleke** [reply, marked as answer]
Hi @remy90,

Assuming this is the requirement:

> The next button should always be enabled when we know there's another question. Ideally, a loading screen is only displayed because of large data fetches or slow network requests, so the first option seems best IMO

I would do it like this:

1. If I understand you correctly, you know the total number of slides in advance. So create empty slides for all questions when you declare your `useState()` for the `slides`:

```tsx
interface QuestionType {
  hasLoaded: boolean
  isLoading: boolean
  questionDetails: {
    title: string
    // ...additional data on this model
  }
}

function emptyQuestionsFromNumber(questionCount: number): QuestionType[] {
  const questions = [...Array(questionCount).keys()]
  return questions.map(() => {
    hasLoaded: false,
    isLoading: false,
    questionDetails: {
      title: ''
      // ...additional data on this model
    }
  })
}

const EmblaCarousel: React.FC<PropType> = (props) => {
  const [slides, setSlides] = useState(emptyQuestionsFromNumber(10))

  // ...

  return (
    <div className="embla">
      <div className="embla__viewport" ref={emblaRef}>
        <div className="embla__container">
          {slides.map((slide, index) => (
            <div className="embla__slide" key={index}>
              {/* Question content... */}
            </div>
          ))}
        </div>
      </div>

      <PrevButton onClick={scrollPrev} enabled={prevBtnEnabled} />
      <NextButton onClick={scrollNext} enabled={nextBtnEnabled} />
    </div>
  )
}
```

2. And then, I would listen to the [`select`](https://www.embla-carousel.com/api/events/#select) event because it will trigger as soon as the user either clicks the next or previous button, or when the user drags the carousel and a selected snap change occur:

```ts
const loadQuestion = useCallback(async (index: number) => {
  // ... load question here
  const questionResponseData = await loadSlide(index)

  // ...and when loading has finished update the slide state
  setSlides(currentSlides => {
    const newSlides = [...currentSlides]
    newSlides[index] = questionResponseData
    return newSlides
  })
}, [])

useEffect(() => {
    if (!emblaApi) return

    // Load first question
    const selectedIndex = emblaApi.selectedScrollSnap()
    loadQuestion(selectedIndex)

    // Load the rest when the selected snap changes
    emblaApi.on('select', () => {
      const selectedIndex = emblaApi.selectedScrollSnap()
      const { isLoading, hasLoaded } = slides[selectedIndex]
      const shouldLoadQuestion = !isLoading && !hasLoaded

      if (shouldLoadQuestion) loadQuestion(selectedIndex)
    })
}, [emblaApi, slides, loadQuestion])
```

Based on the information I have so far, I don't think it has to be more complicated than that. I haven't tested the code above and there might be some things you need to tweak to get it working. But you should get the idea. 

Best,
David

**Reply by davidjerleke** [reply]
@remy90 have you had the chance to read [my comment](https://github.com/davidjerleke/embla-carousel/discussions/467#discussioncomment-5815373)?

**Reply by remy90** [reply]
Hi David, thanks for your response. I removed 'on settle', and it simply worked as expected with smooth transitions. Not sure what I changed, it was just one of those things 🤷‍♂️

Comparing your implementation with mine, I think it's pretty much the same, without on settle. I can now remove my local useState for setSelectedIndex too:
```typescript
  useEffect(() => {
    setEmbla({
      emblaApi: emblaCarouselApi,
      emblaRef: emblaCarouselRef,
    })
    setTotalQuestions(totalQuestionsNo)
    setQuestions([questionSets])
  }, [emblaCarouselApi, emblaCarouselRef, setEmbla])

  useEffect(() => {
    if (emblaApi?.reInit) {
      emblaApi?.reInit()
    }
  }, [emblaApi])

  const onSelect = useCallback(async () => {
    if (!emblaApi) return

    setPrevBtnEnabled(emblaApi.canScrollPrev())
    setNextBtnEnabled(emblaApi.canScrollNext())
  }, [emblaApi])

  useEffect(() => {
    if (!emblaApi) return

    onSelect()
    emblaApi.on('select', onSelect)
    emblaApi.on('reInit', onSelect)
  }, [emblaApi, onSelect])

  return  <div className={classes.embla} ref={emblaRef}>
      <div className={classes.emblaContainer}>
          {Array.from(Array(totalQuestions).keys())?.map(questionIndex => (
          <div className={classes.emblaSlide} key={questionIndex}>
            <div className={classes.pageBody}>
              {questions[questionIndex] ? (
                <QuestionCard
                  question={questions[questionIndex]}
                  onAnswerUpdate={handleAnswerUpdate}
                />
              ) : ( <Loading /> )
              }
              <div className={classes.bodyFooter}>
                <div className={classes.bodyFooterContainer}>
                    <Button  label='back' onClick={handlePrev} />
                    <Button label='save' onClick={handleSubmit} />
                    <Button label='skip' onClick={handleNext} />
                </div>
              </div>
            </div>
          </div>
        ))}
```
I was also thinking of a refactor at some stage to use vanilla, as the whole carousel can be moved to a react server component, with the exception of the navigation buttons. Sounds like there's two potential sandbox demo's (unknown data and server components) that could be created, if you think either of them are of value, I may pick up one of them when I have a little more bandwidth!
